### PR TITLE
Fix casting

### PR DESCRIPTION
**Pull request by pataei**
This PR addresses some of the issues mentioned in #1223. Specifically, it addresses the following:

- Uses consistent terminology to talk about the implicit cast
- Expands explicit cast enumeration
- Adds comments in some examples where implicit cast occurs
- Adds the recursiveness of casting

Some questions arose while formalizing some of P4 which are related to this PR:
- [x] Should we add record types (or anonymous header types) so that we can explicitly add them to the casting section, instead of dancing around it every time we use record types to initialize a struct or header? Discussed this in the LDWG meeting on March 6th. Not sure if there's a consensus. Suggestion: propose the change in a PR. 
- [x] Other than the potential casting of the subtypes of a list, can a list be casted to another type or be the result of a cast? No, it seems that the value of a list can only be written using an explicit cast expression, however, I don't think spec defines it as such. It seems that that's just part of the syntax.
- [x] Are there any restrictions for the nesting of generic types? No, this is checked when types are instantiated.

### CODE REVIEW

File: p4-16/spec/P4-16-spec.mdk [hunk 1]
```diff
@@ -1797,7 +1797,7 @@ restricted places in P4 programs.
 
 ### The error type
 
-The error type contains opaque values that can be used to signal
+The error type contains opaque distinct values that can be used to signal
```
Comment (jnfoster) [inline p4-16/spec/P4-16-spec.mdk:1800]: This might be an improvement to the spec, but it seems unrelated to casting?

Comment (pataei) [reply p4-16/spec/P4-16-spec.mdk:1800]: Yes, it's not related to casting. I just fixed little things that came up as I was working on the formalization.

**Comment by pataei** [comment]
I'll break this down into multiple PRs and then submit them.

**Comment by jnfoster** [comment]
Thanks. That'll be much easier to review.
